### PR TITLE
added jax bijector bridge

### PR DESCRIPTION
**Pull request by jonkhler**
@Olllom this is the JAX bijector bridge.

lmkwyt

### CODE REVIEW

**Comment by Olllom** [comment]
Great!! 

Please make jax and jax2torch optional dependencies: 
- add them as optional dependencies to README.md
- replace module-level `imports`
- add a `pytestmark = pytest.mark.skipif(...)` to the test file to skip the tests if jax or jax2torch are missing
- add both to `devtools/condaenv.yml` so that part of the test suit runs with them installed and part of the test suite skips the tests

**Comment by jonkhler** [comment]
Can this be merged?

File: bgflow/nn/flow/transformer/jax.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+
+def affine_transform(x, a, b):
+    """Affine transform."""
+    return x * jnp.exp(a) + b
+
+
+def smooth_ramp(x, power=1, eps=1e-8):
+    """Smooth ramp."""
+    assert power > 0
```
Comment (Olllom) [inline bgflow/nn/flow/transformer/jax.py:18]: `assert isinstance(power, int)`

File: bgflow/nn/flow/transformer/jax_bridge.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+except ImportError:
+    jax = None
+    jnp = None
+    jax2torch = None
+import functools
+
+
+def apply_r(f, g):
+    """Function application from right."""
+    # (A -> B) (A -> B -> C) => (A -> C)
```
Comment (Olllom) [inline bgflow/nn/flow/transformer/jax_bridge.py:14]: better doc

File: bgflow/nn/flow/transformer/jax_bridge.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+def bisect(bijector, left_bound, right_bound, eps=1e-6):
+    """Bisection search."""
+
+    @jax.jit
+    def _inverted(target):
+        init = (jnp.ones_like(target) * left_bound,
+                jnp.ones_like(target) * right_bound)
+        n_iters = jnp.ceil(-jnp.log2(eps)).astype(int)
+
+        def _body(_, val):
```
Comment (Olllom) [inline bgflow/nn/flow/transformer/jax_bridge.py:42]: val = left_right

File: bgflow/nn/flow/transformer/jax_bridge.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    """Inverts a bijector with a root finder
+       and computes correct gradients using
+       implicit differentation."""
+
+    def _forward(outp, cond):
+        root = root_finder(lambda x: bijector(x, cond)[0])(outp)
+        _, ldj = bijector(root, cond)
+        return (root, -ldj), (root, cond)
+
+    def _backward(res, tangents):
```
Comment (Olllom) [inline bgflow/nn/flow/transformer/jax_bridge.py:65]: to test

File: bgflow/nn/flow/transformer/jax_bridge.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+            left_bound = jnp.where(pred < target, cand, left_bound)
+            right_bound = jnp.where(pred > target, cand, right_bound)
+            return left_bound, right_bound
+
+        return jax.lax.fori_loop(0, n_iters, _body, init)[0]
+
+    return _inverted
+
+
+def invert(bijector, root_finder):
```
Comment (Olllom) [inline bgflow/nn/flow/transformer/jax_bridge.py:55]: rename `invert bijector`

File: bgflow/nn/flow/transformer/jax_bridge.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    def _forward(x, cond):
+        return bijector(x, *cond)
+
+    def _inverse(x, cond):
+        return invert_(_forward)(x, cond)
+
+    return _forward, _inverse
+
+
+def to_bgflow(bijector, domain=None):
```
Comment (Olllom) [inline bgflow/nn/flow/transformer/jax_bridge.py:140]: rename `to_torch`

Comment (Olllom) [reply bgflow/nn/flow/transformer/jax_bridge.py:140]: `to_bgflow`: use forward and inverse in a `Transformer`

File: tests/nn/flow/transformer/test_jax_bridge.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    remap_to_unit,
+    smooth_ramp
+)
+
+
+@pytest.mark.skipif(jax is None,
+                    reason='skipping test due missing jax installation')
+@pytest.mark.skipif(jax2torch is None,
+                    reason='skipping test due to missing jax2torch installation')
+def test_simple_mixture_transformer():
```
Comment (Olllom) [inline tests/nn/flow/transformer/test_jax_bridge.py:33]: test on gpu and cpu as well as float32 and float64

**Review by Olllom — CHANGES_REQUESTED** [review]

File: bgflow/nn/flow/transformer/jax_bridge.py (hunk 6 above)
Comment (Olllom) [reply bgflow/nn/flow/transformer/jax_bridge.py:140]: can we write an autodoc tool to parse parameters?

File: bgflow/nn/flow/transformer/jax_bridge.py [hunk 8]
```diff
@@ -118,31 +125,123 @@ def bijector_with_approx_inverse(bijector, domain=None):
...
+def jax_compile(bijector, vmap_indices, backend, domain=None,):
     """Wraps simple JAX bijector into a transformer,
        that can be used within the bgflow eco-system."""
-    return map(compose(
-        functools.wraps(bijector),
-        jax2torch.jax2torch,
-        jax.jit,
-        jax.vmap),
-        bijector_with_approx_inverse(bijector, domain))
+    compile_bijector = compose(functools.partial(jax.jit, backend=backend), wrap_params)
```
Comment (Olllom) [inline bgflow/nn/flow/transformer/jax_bridge.py:176]: nice! 
Now that you have a dedicated function and caching for this, it should be easy to support double precision, too.

**Comment by jonkhler** [comment]
There is still an issue with the gradients - currently looking into it.

File: bgflow/nn/flow/transformer/jax_bridge.py (hunk 8 above)
Comment (jonkhler) [reply bgflow/nn/flow/transformer/jax_bridge.py:176]: Very well possible :)

**Review by Olllom — APPROVED** [review]

**Comment by jonkhler** [comment]
TODO(@jonkhler): 
- unit test for correct gradients
- autodoc
- make work for @Olllom setup

**Comment by Olllom** [comment]
Note: Using the bridge on a GPU requires a python environment that supports jax, torch, and openmm on CUDA. That's not trivial setting up. We should add some documentation on how to do this.
